libCacheSim/cache/prefetch/PG.c

```diff
@@ -0,0 +1,380 @@
+
+  PG_params->stop_recording = FALSE;
+
+  PG_params->graph = g_hash_table_new_full(g_direct_hash, g_direct_equal, NULL, _graphNode_destroy);
+  PG_params->prefetched = g_hash_table_new_full(g_direct_hash, g_direct_equal, NULL, NULL);
+  PG_params->past_requests = g_new0(guint64, PG_params->lookahead_range);
+
+  PG_params->past_request_pointer = 0;
+  PG_params->num_of_hit = 0;
+  PG_params->num_of_prefetch = 0;
+
+  PG_params->cache_size_map = g_hash_table_new_full(g_direct_hash, g_direct_equal, NULL, NULL);
+}
+
+// ***********************************************************************
+// ****                                                               ****
+// ****                     prefetcher interfaces                     ****
+// ****                                                               ****
+// ****   create, free, clone, handle_find, handle_evict, prefetch    ****
+// ***********************************************************************
+prefetcher_t *create_PG_prefetcher(const char *init_params, uint64_t cache_size);
+/**
+ 1. record the request in cache_size_map for being aware of prefetching object's
+ size in the future.
+ 2. call `_PG_add_to_graph` to update graph.
+
+ @param cache the cache struct
+ @param req the request containing the request
+ @return
+*/
+static void PG_handle_find(cache_t *cache, const request_t *req, bool hit) {
+  PG_params_t *PG_params = (PG_params_t *)(cache->prefetcher->params);
+
+  /*use cache_size_map to record the current requested obj's size*/
+  g_hash_table_insert(PG_params->cache_size_map, GINT_TO_POINTER(req->obj_id), GINT_TO_POINTER(req->obj_size));
+
+  _PG_add_to_graph(cache, req);
+
+  if (g_hash_table_contains(PG_params->prefetched, GINT_TO_POINTER(req->obj_id))) {
+    PG_params->num_of_hit++;
+    g_hash_table_remove(PG_params->prefetched, GINT_TO_POINTER(req->obj_id));
+    if (g_hash_table_contains(PG_params->prefetched, GINT_TO_POINTER(req->obj_id))) {
+      fprintf(stderr, "ERROR found prefetch\n");
+    }
+  }
+}
+
+/**
+ remove this obj from `prefetched` if it was previously prefetched into cache.
+
+ @param cache the cache struct
+ @param req the request containing the request
+ @return
+*/
+void PG_handle_evict(cache_t *cache, const request_t *check_req) {
+  PG_params_t *PG_params = (PG_params_t *)(cache->prefetcher->params);
+
+  g_hash_table_remove(PG_params->prefetched, GINT_TO_POINTER(check_req->obj_id));
+}
+
+/**
+ prefetch some objects which are from `_PG_get_prefetch_list`
+
+ @param cache the cache struct
+ @param req the request containing the request
+ @return
+ */
+void PG_prefetch(cache_t *cache, const request_t *req) {
+  PG_params_t *PG_params = (PG_params_t *)(cache->prefetcher->params);
+
+  // begin prefetching
+  GList *prefetch_list = _PG_get_prefetch_list(cache, req);
+  if (prefetch_list) {
+    GList *node = prefetch_list;
+    request_t *new_req = my_malloc(request_t);
+    memcpy(new_req, req, sizeof(request_t));
```

Comment:
it might be better to use `copy_request` if we plan to add some tracking when a request is copied 
https://github.com/1a1a11a/libCacheSim/blob/3dc3f678722b6d89ae8eb3e1699ad81a8679b060/libCacheSim/include/libCacheSim/request.h#L90

Reply:
good suggestion!